feat: Add SBOM generation workflow using Anchore SBOM Action

peatey opened the pull request:
This commit introduces a new GitHub Actions workflow for generating Software Bill of Materials (SBOM) using the Anchore SBOM Action.

Key features:
- Generates SBOM for source code in both SPDX and CycloneDX formats
- Generates SBOM for container images on releases
- Automatically attaches SBOMs to GitHub releases
- Runs on release tags and can be manually triggered
- Includes PR validation for source code SBOMs

The workflow provides comprehensive visibility into project dependencies and supply chain security, supporting both industry-standard SBOM formats.

## What does this PR change?
* uses anchore/sbom-action@v0 to attach an SBOM to our releases. 

## Does this PR relate to any other PRs?
* no

## How will this PR impact users?
* invisible - it simply adds the SBOM

## Does this PR address any GitHub or Zendesk issues?
* Closes ...

## How was this PR tested?
* simply GHA import tested

## Does this PR require changes to documentation?
* that is a good point. Short of the creation of the SBOM and its attachment to the release no thought has been given to a singular storage and presentation location

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
*

claude committed:
feat: Add SBOM generation workflow using Anchore SBOM Action

This commit introduces a new GitHub Actions workflow for generating
Software Bill of Materials (SBOM) using the Anchore SBOM Action.

Key features:
- Generates SBOM for source code in both SPDX and CycloneDX formats
- Generates SBOM for container images on releases
- Automatically attaches SBOMs to GitHub releases
- Runs on release tags and can be manually triggered
- Includes PR validation for source code SBOMs

The workflow provides comprehensive visibility into project dependencies
and supply chain security, supporting both industry-standard SBOM formats.

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,157 @@
+          RELEASE_VERSION: ${{ steps.version_number.outputs.RELEASE_VERSION }}
+        run: |
+          if [ "${{ github.event_name }}" == "pull_request" ]; then
+            echo "IMAGE_TAG=ghcr.io/$REPO_OWNER/opencost:develop-latest" >> $GITHUB_OUTPUT
```
on a PR, should we use the container image built for that branch? an SBOM scan on develop won't tell us anything about the branched container. Alternatively - open to changing this to not run on PR and only run on develop merges and releases

peatey replied:
updated to only run on develop merges and releases

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,157 @@
+          output-file: opencost-container-sbom.cyclonedx.json
+          format: cyclonedx-json
+          registry-username: ${{ github.actor }}
+          registry-password: ${{ secrets.GITHUB_TOKEN }}
```
I don't think you need these? all our images are world readable

peatey replied:
updated to remove

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,157 @@
+        if: startsWith(github.ref, 'refs/tags/')
+        uses: anchore/sbom-action/publish-sbom@v0
+        with:
+          sbom-artifact-match: ".*\\.spdx\\.json$|.*\\.cyclonedx\\.json$"
```
this action will automatically know to attach these sboms to the right release? should we test?

peatey replied:
thats correct we can run the test manually against any release currently sitting

peatey replied:
Yes, the anchore/sbom-action/publish-sbom action automatically detects the GitHub release associated with the tag that triggered the workflow and attaches the SBOMs to it. This is a built-in feature of the action.

We can verify this works on the next release, or we can manually test using workflow_dispatch after this PR is merged. The action uses the github.ref context to find the correct release.

claude committed:
fix: Remove unnecessary image tag logic for PR workflows

Addresses reviewer feedback from @ameijer. The workflow already skips
container image SBOM generation on PRs (only scans source code), but
the image tag step was still running and setting an unused value.

Changes:
- Add condition to skip "Set OpenCost Image Tag" step on PRs
- Remove dead PR-specific logic since IMAGE_TAG is never used on PRs
- Clarifies that PRs only generate source code SBOMs

On PRs: Only source code SBOMs are generated
On releases: Both source code and container image SBOMs are generated

claude committed:
fix: Remove unnecessary registry authentication for public images

Addresses reviewer feedback from @ameijer. Since OpenCost container
images are publicly accessible, there's no need to authenticate to
the registry when scanning them for SBOM generation.

Changes:
- Removed registry login step
- Removed registry-username and registry-password parameters from
  both container image SBOM generation steps (SPDX and CycloneDX)

This simplifies the workflow and removes unnecessary credential usage.

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,142 @@
+        if: github.event_name != 'pull_request'
+        uses: anchore/sbom-action@v0
+        with:
+          image: ${{ steps.image_tag.outputs.IMAGE_TAG }}
```
assuming this runs in parallel with release image generation, this logic is going to have to wait for the image to get built and become pullable

claude committed:
fix: Add workflow dependency to wait for image build completion

Addresses reviewer feedback from @ameijer. The SBOM workflow now waits
for the "Build and Publish Release" workflow to complete before running,
ensuring the container image is built and published before we attempt
to scan it.

Changes:
- Replaced push:tags trigger with workflow_run trigger
- Workflow now triggers after "Build and Publish Release" completes
- Added condition to only run if build workflow succeeded
- Updated tag extraction logic to work with workflow_run events
- Updated publish condition to work with new trigger mechanism

This prevents race conditions where the SBOM workflow tries to pull
an image that hasn't been published yet.

Workflow triggers:
- workflow_run: After successful release build (for releases)
- workflow_dispatch: Manual trigger (for testing/re-runs)
- pull_request: PR validation (source code only)

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,146 @@
+          path: .
+          artifact-name: opencost-source-sbom.cyclonedx.json
+          output-file: opencost-source-sbom.cyclonedx.json
+          format: cyclonedx-json
```
let's at least cat the generated sbom on PRs so that we don't just generate them for no reason

peatey replied:
Added a step to display SBOM contents in the GitHub Actions summary for PRs. Reviewers will now see:

Total package count
List of first 50 packages with versions
Full SBOM in a collapsible section
This makes the PR SBOM generation actually useful for review. Changes pushed in commit aad57e7.

claude committed:
feat: Display SBOM contents in PR summary for review

Addresses reviewer feedback from @ameijer. On PRs, the workflow now
displays the generated SBOM contents in the GitHub Actions summary
so reviewers can see what dependencies are included.

Changes:
- Added "Display SBOM Contents" step for PRs only
- Shows total package count
- Lists first 50 packages with versions
- Provides full SBOM in collapsible section for detailed review

This makes PR SBOM generation useful by giving visibility into
dependency changes without having to download artifacts.

claude committed:
fix: Address CodeQL security alert for untrusted checkout

Fixes CodeQL alert "Potential execution of untrusted code on a
privileged workflow" by implementing the principle of least privilege.

Changes:
- Set top-level permissions to read-all (most restrictive by default)
- Made contents permission conditional:
  - Read-only for pull_request events (prevents untrusted code execution)
  - Write permission only for releases (needed to attach SBOMs)
- Simplified checkout to use default behavior for PRs instead of
  explicitly checking out head_ref (safer)

This ensures PRs from forks cannot execute with write permissions,
mitigating the security risk while maintaining functionality for
release SBOM publishing.

Security: Addresses actions/untrusted-checkout/high

ameijer committed:
Merge branch 'develop' into claude/secure-bill-of-materials-011CV4L7f7X28t8xeuLGhTFz

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,167 @@
+        run: |
+          echo "TRIGGERED_TAG=${GITHUB_EVENT_WORKFLOW_RUN_HEAD_BRANCH#refs/*/}" >> $GITHUB_ENV
+        env:
+          GITHUB_EVENT_WORKFLOW_RUN_HEAD_BRANCH: ${{ github.event.workflow_run.head_branch }}
```
last q - don't we want to grab the tag at the head of the branch?

ameijer commented on .github/workflows/sbom.yml:
```diff
@@ -0,0 +1,167 @@
+      - name: Determine Version Number
+        id: version_number
+        env:
+          RELEASE_VERSION: ${{ inputs.release_version }}
```
this won't be set when being triggered from upstream workflow

peatey replied:
courtesy of our friend - https://github.com/opencost/opencost/pull/3455/commits/bce07f3b9c478f794aec710bc2ef24737a5b9979

claude committed:
fix: Properly detect version from workflow_run events

Addresses reviewer feedback from @ameijer. When triggered by workflow_run,
inputs.release_version is not available, so we need to extract the version
from the workflow run event data.

Changes:
- Added early checkout for workflow_run to enable git operations
- Use git to find the tag pointing to the workflow run's head SHA
- Simplified version determination logic with explicit event type checks
- Avoid redundant checkout by skipping second checkout for workflow_run

The workflow now correctly determines the release version regardless of
how it's triggered:
- workflow_run: Finds tag from commit SHA
- workflow_dispatch: Uses inputs.release_version
- pull_request: No version needed (source code only)

ameijer reviewed (approved):
LGTM
